routing/datasource.go

```diff
@@ -585,6 +586,23 @@ func receiveRouteMatcher(o Options, out chan<- *routeTable, quit <-chan struct{}
 			updatesRelay = updates
 			outRelay = nil
 		case <-quit:
+			// cleanup with postprocessors
+			for i := range o.PreProcessors {
+				defs = o.PreProcessors[i].Do(defs)
+			}
+			routes, _ := processRouteDefs(o, o.FilterRegistry, defs)
+			for i := range o.PostProcessors {
+				routes = o.PostProcessors[i].Do(routes)
+			}
+			for _, r := range routes {
+				for _, f := range r.Filters {
+					fc, ok := f.Filter.(filters.FilterCloser)
+					if ok {
+						fc.Close()
+					}
+				}
+			}
+
```

Comment:
I guess we can just keep a pointer to the routes:
```diff
diff --git a/routing/datasource.go b/routing/datasource.go
index 9b679df4..c2ab4ace 100644
--- a/routing/datasource.go
+++ b/routing/datasource.go
@@ -520,6 +520,7 @@ func processRouteDefs(o Options, fr filters.Registry, defs []*eskip.Route) (rout
 
 type routeTable struct {
        m             *matcher
+       routes        []*Route
        validRoutes   []*eskip.Route
        invalidRoutes []*eskip.Route
        created       time.Time
@@ -574,6 +575,7 @@ func receiveRouteMatcher(o Options, out chan<- *routeTable, quit <-chan struct{}
 
                        rt = &routeTable{
                                m:             m,
+                               routes:        routes,
                                validRoutes:   validRoutes,
                                invalidRoutes: invalidRoutes,
                                created:       time.Now().UTC(),
@@ -585,6 +587,13 @@ func receiveRouteMatcher(o Options, out chan<- *routeTable, quit <-chan struct{}
                        updatesRelay = updates
                        outRelay = nil
                case <-quit:
+                       for _, route := range rt.routes {
+                               for _, f := range route.Filters {
+                                       if fc, ok := f.Filter.(filters.FilterCloser); ok {
+                                               fc.Close()
+                                       }
+                               }
+                       }
                        return
                }
        }
```

Reply:
I think we can do that in the startReceivingUpdates

```go
--- a/routing/routing.go
+++ b/routing/routing.go
@@ -371,6 +371,10 @@ func (r *Routing) startReceivingUpdates(o Options) {
                                }
                                r.log.Info("route settings applied")
                        case <-r.quit:
+                               rt := r.routeTable.Load().(*routeTable)
+                               if rt != nil {
+                                       // ...
+                               }
                                return
                        }
                }
```

Reply:
The problem with routeTable is that we do not have all routes here, only access to leafRoutes as far as I tried last week.

Reply:
I guess you suggest to add the routes to the routeTable to store the link to all the routes, which would be good enough to fix the problem, yes.

Reply:
yeah, keep the link and close filters from startReceivingUpdates

Reply:
Thanks for the input I put it into Routing.Close, because there it makes most sense and then we do not have the problem of nil ptr.
Let's see what the tests say.